x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/Expression.java

```diff
@@ -29,6 +29,32 @@
  */
 public abstract class Expression extends Node<Expression> implements Resolvable {
 
+    public enum Nullable {
+        POSSIBLY,  // Whether the expression becomes null if at least one param/input is null
+        NEVER,     // The expression can never become null
+        UNKNOWN;   // Cannot determine if the expression supports possible null folding
+
+        public static Nullable and(Nullable... args) {
+            if (args.length == 0) {
+                return UNKNOWN;
+            }
+
+            Nullable returnValue = args[0];
+
+            // UKNOWN AND <anything> => UKNOWN
+            // NEVER AND NEVER => NEVER
+            // POSSIBLE AND NEVER/POSSIBLE => POSSIBLE
+            for (int i = 1; i < args.length; i++) {
```

Comment:
The `NEVER` case doesn't seem to be handled. Also it might make sense to assign some bitmasks to the enum to make the comparison simpler and use just one accumulator.
That is, instead of checking the return value and the next value, it would be easier to just 'combine' the current value (returnValue) with the next value and be done with.
(an OO way would be to define a method on the enum but that adds a virtual call for minimum gain).
Further more since UNKNOWN trumps everything, the loop could be skipped if this value is found:

```java

Nullability value = null;
for (Nullability n: nullables) {
    switch (n) {
       case UNKNOWN: return UNKNOWN;
       case POSSIBLE: value = n;
       case NEVER: 
          if (value == null) {
              value = n;
          }
     }
}

return value;
```

Reply:
I like the bitmask idea!